### PR TITLE
Custom translation for `random()`

### LINKED ISSUE #617: slice_sample() throwing error on a MariaDB database
I've recently had to pull a random sample of a fairly large (>1.5M row) database that's stored in a MariaDB instance. I wrote a fairly basic dbplyr filter:

```r
# pseudocode: <set up database connection>
# pseudocode: db_tbl <- tbl( <connection>, "tablename" )

all_articles <-   
   db_tbl %>%
   slice_sample( n=5000 ) %>%
   collect 

```

I receive:

```
Error: FUNCTION tablename.random does not exist [1305]
```

I assume that this may be a limitation in functions internally provided by MariaDB, but I'm also unsure from searching around whether this is something I can fix (by implementing tablename.random directly on the DB side?) or if I'm just missing something obvious.

As an aside, I'm in a lucky position here where we're about to move this whole dataset and collection across to a new system, so I have the option to move database servers. Is there a recommended database implementation that's most compatible for dbplyr?

### LINKED ISSUE #650: slice_sample doesn't work with microsoft sql
tried to produce samples in groups. 

Got this error below

```r
camp_tbl <- tbl(con,dbplyr::in_schema('schema','campaigns'))

unique_df <- 
  camp_tbl %>% 
  rename_all(.,.funs = tolower) %>% 
  select(campaign,starts_with('item')) %>% 
  group_by(campaign) %>%
  slice_sample(n = 5) %>% 
  distinct() %>% 
  collect()
```
Error message below
```
Error: nanodbc/nanodbc.cpp:1655: 42000: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]'random' is not a recognized built-in function name.  [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Incorrect syntax near 'q01'.  [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Statement(s) could not be prepared.
```

version
```
> sessionInfo()
R version 4.0.5 (2021-03-31)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_Australia.1252  LC_CTYPE=English_Australia.1252   
[3] LC_MONETARY=English_Australia.1252 LC_NUMERIC=C                      
[5] LC_TIME=English_Australia.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_1.0.5  odbc_1.3.2   dbplyr_2.1.1 tibble_3.1.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6       rstudioapi_0.13  magrittr_2.0.1   hms_1.0.0        tidyselect_1.1.0
 [6] bit_4.0.4        R6_2.5.0         rlang_0.4.10     fansi_0.4.2      blob_1.2.1      
[11] tools_4.0.5      xfun_0.22        tinytex_0.31     utf8_1.2.1       cli_2.4.0       
[16] DBI_1.1.1        withr_2.4.1      ellipsis_0.3.1   bit64_4.0.5      assertthat_0.2.1
[21] lifecycle_1.0.0  crayon_1.4.1     purrr_0.3.4      vctrs_0.3.7      glue_1.4.2      
[26] compiler_4.0.5   pillar_1.6.0     generics_0.1.0   pkgconfig_2.0.3 
```

### PR DESCRIPTION
**Pull request by mgirlich**
Fixes #617 and fixes #650

### CODE REVIEW

**Review by krlmlr — COMMENTED** [review]
As this is internal to dbplyr, do we really need to add a new translation? Would a new `sql_rand()` generic be overkill here?

If we keep the translation, to avoid potential for confusion, should we declare and use the function as `RANDOM()` ?

**Comment by mgirlich** [comment]
So far, the translations are focused on existing R functions. Therefore it might make sense to not add this as a new translation.

File: R/backend-mysql.R [hunk 1]
```diff
@@ -131,5 +131,14 @@ sql_expr_matches.MySQL <- sql_expr_matches.MariaDBConnection
...
 sql_expr_matches.MySQLConnection <- sql_expr_matches.MariaDBConnection
 
+#' @export
+sql_random.MariaDBConnection <- function(con) {
+  sql_expr(RAND())
+}
+
+#' @export
+sql_random.MySQLConnection <- sql_random.MariaDBConnection
+sql_random.MySQL <- sql_random.MariaDBConnection
```
Comment (hadley) [inline R/backend-mysql.R:141]: Missing `@export`

**Review by hadley — APPROVED** [review]

**Comment by HarlanH** [comment]
FYI, this appears to be an issue with BigQuery too. Should I open a new issue for that?

```
> slice_sample(dat_window2, n=10000) %>% show_query()
<SQL>
SELECT `days_to_review`, `satisfaction_score_numeric`
FROM (
  SELECT
    `days_to_review`,
    `satisfaction_score_numeric`,
    ROW_NUMBER() OVER (ORDER BY RANDOM()) AS `q01`
  FROM (
    SELECT `days_to_review`, `satisfaction_score_numeric`
    FROM `tpt-data-warehouse-prod.dev_harlan_harris.ratings_over_time`
    WHERE (NOT(`is_free`)) AND (`acquire_ts` <= '2021-07-02 09:57:21')
  ) `q01`
) `q02`
WHERE (`q01` <= 10000)
```

That should be `ORDER BY RAND()`...

**Comment by hadley** [comment]
@HarlanH yes please, just so we don't lose it.

**Pull request merged**